Implement changes to Cursor API to include the stream transaction id …

DiscJockeyDJ opened the pull request:
…to the request headers.

fix #322

DiscoPYF commented on arangodb-net-standard/CursorApi/CursorApiClient.cs:
```diff
@@ -90,7 +112,8 @@ public CursorApiClient(IApiClientTransport client, IApiClientSerialization seria
             var content = GetContent(postCursorBody, new ApiClientSerializationOptions(true, true));
-            using (var response = await _client.PostAsync(_cursorApiPath, content).ConfigureAwait(false))
+            var headerCollection = GetHeaderCollection(postCursorBody);
+            using (var response = await _client.PostAsync(_cursorApiPath, content, headerCollection))
```
`ConfigureAwait(false)` should be preserved, same for other endpoints below.

DiscoPYF reviewed (commented):
The current implementation looks good.

A `transactionId` parameter has been added to `DeleteCursorAsync`, `PutCursorAsync` and `PostCursorBody`, but I wonder if we could instead add an additional general parameter to the API methods for supporting headers, like we do for query/body parameters? e.g. `PostCursorAsync<T>(PostCursorBody postCursorBody, PostCursorHeaders postCursorHeaders)`, or something a bit more generic.

https://www.arangodb.com/docs/stable/http/transaction-stream-transaction.html
> To use the transaction for a supported operation a client needs to specify the transaction identifier in the x-arango-trx-id header on each request.

I think it's worth taking this opportunity to model something, like an interface containing common headers.

@rossmills99 @modios Any thoughts?

There are headers that are applied widely, e.g. `x-arango-async: true`, but I'm not sure it's a strong enough reason to add support for them on each endpoint, as they could also be implemented differently.

rossmills99 commented:
I notice from [the docs](https://www.arangodb.com/docs/stable/http/transaction-stream-transaction.html) that the transaction ID header is only relevant in the "create cursor" endpoint, so our DeleteCursor or PutCursor APIs should not be affected.  Those changes should be removed from the PR.

To address @DiscoPYF 's questions:

> I wonder if we could instead add an additional general parameter to the API methods for supporting headers, like we do for query/body parameters? e.g. `PostCursorAsync<T>(PostCursorBody postCursorBody, PostCursorHeaders postCursorHeaders)`, or something a bit more generic.

I think that there are two PostCursorAsync methods.  One of them is a convenience method that provides some of the options are method arguments, instead of requiring the `PostCursorBody` object.  I think that it would make sense to include `string transactionId = null` as an optional argument in the convenience method, and then for the other method introduce a new type that allows additional headers more generically in the request.  At least I think that would be consistent with how we've approached things so far.  What do you think @DiscoPYF ?

DiscoPYF commented:
> I think that it would make sense to include string transactionId = null as an optional argument in the convenience method, and then for the other method introduce a new type that allows additional headers more generically in the request.

I agree, this is a good approach (imo) and consistent with our API, indeed.

DiscJockeyDJ commented:
> then for the other method introduce a new type that allows additional headers more generically in the request
> Example: WebHeaderCollection vs a model with common header properties.

Instead of creating a new `TransactionId` property in _PostCursorBody_, should a new model be created ? Is that correct ?

DiscoPYF commented:
Yes, correct. I think the idea is to have the following:

```c#
public virtual async Task<CursorResponse<T>> PostCursorAsync<T>(
            PostCursorBody postCursorBody, AClassWithHeaderProperties headers)
```

DiscJockeyDJ commented:
PR updated addressing these comments.

> DeleteCursor or PutCursor APIs should not be affected.

> Instead of creating a new TransactionId property in PostCursorBody, should a new model be created? Is that correct?
> Yes, correct

rossmills99 commented:
Thanks for the update. It appears `ConfigureAwait` was removed from a lot of the async calls.  Can you replace it please?  This was added for reasons described in #215

DiscJockeyDJ commented:
PR updated.

Reinstated the missed out `ConfigureAwait` to the endpoints.

DiscoPYF reviewed (commented):
@DiscJockeyDJ Thank you for the update. This implementation fits well with the current APIs. I left minor comments about naming and placing of new methods/classes, but this could be adjusted in an upcoming pull request.

I noticed that some test runs are failing. I will take a look, but approved pending failure resolved.

DiscoPYF commented:
It would also be useful to add a test to verify that header properties are being passed on to transport correctly. Example: `PostCursorAsync_ShouldUseHeaderProperties` can call `PostCursorAsync` and mock `IApiClientTransport` to check that the provided `WebHeaderCollection` contains the expected header key/values. Similar to what `PostCollectionAsync_ShouldUseQueryParameter` does.

DiscoPYF commented:
I ran again the failed test runs and they now pass.

I am listing the failures for future reference:

- https://app.circleci.com/pipelines/github/Actify-Inc/arangodb-net-standard/614/workflows/54ac61ec-5caa-41c0-bb96-32aebf07aa25/jobs/1298
```
System.Net.Sockets.SocketException : Connection refused
```
- https://app.circleci.com/pipelines/github/Actify-Inc/arangodb-net-standard/614/workflows/54ac61ec-5caa-41c0-bb96-32aebf07aa25/jobs/1301
```
ArangoDBNetStandard.ApiErrorException : service unavailable due to startup or maintenance mode
```

DiscJockeyDJ commented on arangodb-net-standard/CursorApi/CursorApiClient.cs:
```diff
@@ -48,16 +48,16 @@ public CursorApiClient(IApiClientTransport client, IApiClientSerialization seria
+        /// <param name="headerProperties">The <see cref="HeaderProperties"/> values.</param>
         /// <returns><see cref="WebHeaderCollection"/> values.</returns>
-        private static WebHeaderCollection GetHeaderCollection(PostCursorBody body)
+        private static WebHeaderCollection GetHeaderCollection(HeaderProperties headerProperties)
```
Created this method through _Quick Actions & Refactoring Options_ in VS. Will remove the `static` keyword and make the method `protected virtual` +1

DiscJockeyDJ commented:
PR updated to address the above mentioned suggestions 👍

> If this class is only for the cursor API, maybe we should name it more explicitely? e.g. CursorHeaderProperties

> I am thinking it would be useful to have it protected virtual

> It would also be useful to add a test to verify that header properties are being passed on to transport correctly. Example: PostCursorAsync_ShouldUseHeaderProperties

DiscoPYF reviewed (approved):
Thank you @DiscJockeyDJ , looks good to me. Approved 👍

rossmills99 reviewed (approved):
Looks good to me
